File: rails/app/models/vendor.rb
# frozen_string_literal: true

class Vendor < ApplicationRecord
  include Sluggable
  belongs_to :track
  has_many :menu_items, dependent: :destroy

  def menu
    menu_items.each_with_object(Hash.new { |hash, key| hash[key] = [] }) do |item, by_heading|
      heading = item.heading || '*'
      by_heading[heading] << item.as_json(only: %i[description name slug], methods: :price)
    end
  end
end

File: react-native/src/models/vendor.ts
import { MenuItem } from "./menu-item";

interface BaseVendor {
  name: string;
  slug: string;
}

export interface Vendor extends BaseVendor {
  category: string;
}

export interface VendorWithMenu extends Vendor {
  menu: Record<string, MenuItem[]>;
}

export type CategorizedVendor = BaseVendor;
